pycroft/model/alembic/versions/ff700db83195_add_address_relation.py

```diff
@@ -0,0 +1,244 @@
+    set_uninhabitable_room_addresses(session)
+
+    # DATA MIGRATION IV: set `address_id` to room's address for inhabitable rooms
+    set_inhabitable_room_addresses(session)
+
+    # FURTHER SCHEMA MIGRATION…
+    op.add_column('user', sa.Column('address_id', sa.Integer(), nullable=False,
+                                    server_default=dummy_default_cause))
+    cleanups.append(lambda: op.alter_column('user', 'address_id', server_default=None))
+    op.create_index(op.f('ix_user_address_id'), 'user', ['address_id'], unique=False)
+    op.create_foreign_key(None, 'user', 'address', ['address_id'], ['id'])
+
+    # DATA MIGRATION VI: set `user.address` for users with room
+    set_user_addresses(session)
+
+    # FURTHER SCHEMA MIGRATION (cleanup)
+    for action in cleanups:
+        action()
+
+
+def set_user_addresses(session):
+    session.execute(
+        sa.update(_User)
+        .values(address_id=_RoomAfter.address_id)
+        .where(_RoomAfter.id == _User.room_id)  # implies `room_id` is not null
+    )
+
+
+def set_inhabitable_room_addresses(session):
+    session.execute(
+        sa.update(_Room)
+        .where(_Room.inhabitable)
+        .values(address_id=_Address.id)
+        .where(_Room.id == _Address.tmp_room_id)
+    )
+
+
+def set_uninhabitable_room_addresses(session):
+    session.execute(
+        sa.update(_Room)
+        .where(sa.sql.not_(_Room.inhabitable))
+        .values(address_id=_Address.id)
+        .where(_Room.building_id == _Address.tmp_building_id)
+    )
+
+
+def add_room_addresses(session):
+    room_select = (
+        sa.select([_Room.id, (sa.cast(_Room.level, sa.String)
+                              + "-"
+                              + _Room.number).label("addition")])
+        .select_from(_Room)
+        .select_from(_Building)
+        .where(_Building.id == _Room.building_id)
+        .where(_Room.inhabitable)
+        .column(_Building.street)
+        .column(_Building.number)
+        .alias('inhabitable_room_info')
+    )
+    session.execute(
+        sa.insert(_Address)
+        .from_select([_Address.tmp_room_id, _Address.addition, _Address.street, _Address.number],
+                     room_select)
+    )
+
+
+def add_building_addresses(session):
+    building_select = sa.select([_Building.street, _Building.number, _Building.id])
+    session.execute(
+        sa.insert(_Address)
+        .from_select([_Address.street, _Address.number, _Address.tmp_building_id],
+                     building_select)
+        .returning(_Address.id)
+    )
+
+
+def add_dummy_address(session):
+    # noinspection SpellCheckingInspection
+    dummy_addr = _Address(street="Niemandsstraße", number="42", zip_code="33602",
+                          city="Bielefeld", state="Niedersachsen")
+    session.add(dummy_addr)
+    session.commit()
+    session.refresh(dummy_addr)
+    dummy_id = dummy_addr.id
+    return dummy_id
```

Comment:
Why are we using a dummy address instead of a nullable address_id?

Reply:
And what is the backround of always updating the user `address_id` on room changes instead of keeping it on `null`? The `address` attribute could be a hybrid property returning `address_id` if it is not `null`, `room.address_id` if the user is living in a room or `null` if both conditions weren't met.
The current implementation could lead to inconsistency in the database.

Reply:
See [this comment](https://github.com/agdsn/pycroft/issues/305#issuecomment-537691101).

There's a few questions/decisions which are relevant here.

## When to enforce that every user has an address

At some point in the future, to conform with TKG requirements, we should ensure that **every user has a fully qualified address**. I think it's simpler to already represent this convention in the schema and at a later point in time fix the data, instead of now not requiring the constraint. Because if we add the constraint later, we will have a migration that only works when e.g. you don't have any users without an address information. I think that's not a good idea from an operational perspective.
The cleanup work of deleting the users (see #67) must be done anyway, and whether these users have a `null` address or the dummy address does not matter from a data perspective.

## How to enforce that every user has an address
To ensure this on the sql level, we could 
- Make `user.address_id` non-nullable
- Create a check constraint verifying that every `user` has either an `address_id` or a `room_id`.

Following [@mmarx's comment](https://github.com/agdsn/pycroft/issues/305#issuecomment-469382507), I opted for the first solution due to its simplicity.

## Is the potential inconsistency bad?
For a weak argument, it might even be a use case that someone wants to have a different fully-qualified address than where his room is located.

For a stronger argument, our schema isn't perfectly irredundant, anyway, and relies on consistency by convention because the library functions act „correctly“. For instance, the schema allows a user to have a host without an interface. There's no use case for that.
Explicitly setting an address in just two instances is imo quite a low cost for the simplicity.

For a conceptual argument: The `room` as we use it is actually more of a technical entity: we  use it

- to know which switch ports (also patch ports) are associated to the user (to establish access and diagnostics)
- after that, potentially to identify a person by their room in a support scenario.
This is not necessarily coupled to postal identification.

So i'd argue that from that perspective, it's not incorrect to have the postal information (`user.address_id`) decoupled from the technical location (`user.room_id`) – on the contrary.